spyder/plugins/ipythonconsole.py

```diff
@@ -1355,25 +1360,42 @@ def restart_kernel(self):
             client.restart_kernel()
 
     #------ Public API (for tabs) ---------------------------------------------
-    def add_tab(self, widget, name):
+    def add_tab(self, widget, name, filename=None):
         """Add tab"""
         self.clients.append(widget)
         index = self.tabwidget.addTab(widget, name)
+        self.filenames.insert(index, filename)
         self.tabwidget.setCurrentIndex(index)
         if self.dockwidget and not self.ismaximized:
             self.dockwidget.setVisible(True)
             self.dockwidget.raise_()
         self.activateWindow()
         widget.get_control().setFocus()
+        self.update_tabs_text()
         
     def move_tab(self, index_from, index_to):
         """
         Move tab (tabs themselves have already been moved by the tabwidget)
         """
+        filename = self.filenames.pop(index_from)
         client = self.clients.pop(index_from)
+        self.filenames.insert(index_to, filename)
         self.clients.insert(index_to, client)
         self.update_plugin_title.emit()
 
+    def disambiguate_fname(self, fname):
+        """Generate a file name without ambiguation."""
+        files_path_list = [filename for filename in self.filenames
+                           if filename is not None]

```

Comment:
With my last suggested change, you need to change this line to

    if filename

instead of

    if filename is not None